File: Array/MaxProductSubArray.cpp
#include <vector>
#include <string>
#include <stack>
#include <unordered_map>
#include <set>
#include <iostream>
#include <algorithm>
#include <iterator>

using namespace std;

int maxProduct(vector<int>& nums) {
    int maxProd = nums[0];
    int maxTillHere = nums[0];
    int minTillHere = nums[0];

    for (int i = 1; i < nums.size(); i++) {
        if (nums[i] < 0)
            swap(minTillHere, maxTillHere);

        maxTillHere = max(nums[i], nums[i] * maxTillHere);
        minTillHere = min(nums[i], nums[i] * minTillHere);

        maxProd = max(maxTillHere, maxProd);
    }
    return maxProd;
}
int main() {
    vector<int> nums{3,1,2,-1,4,0, 6,4,-1,3,-2};
    cout<<maxProduct(nums)<<"\n";
    return 0;
}

File: Array/MaxProductSubArray.py
"""
Given an integer array nums, find a contiguous non-empty subarray
within the array that has the largest product, and return the product.
"""

def maxProduct(nums) -> int:
    maxVal = nums[0]
    minVal = nums[0]

    maxProd = nums[0]

    for i in range(1,len(nums)):
        if maxVal == None:
            maxVal = nums[i]
            minVal = nums[i]
        
        if nums[i] < 0:
            maxVal,minVal = minVal, maxVal

        
        maxVal = max(nums[i], maxVal * nums[i])
        minVal = min(nums[i], minVal * nums[i])
        
        maxProd = max(maxProd, maxVal)
    
    return maxProd


if __name__ == '__main__':
    print(maxProduct([1,2,-3,4]))
    print(maxProduct([1,-2,-3,4]))
    print(maxProduct([1,2,0,-3,4]))
    print(maxProduct([1,-2,-8,0,-3,4]))
    print(maxProduct([1,-2,-8,0,-3,4,-9]))
